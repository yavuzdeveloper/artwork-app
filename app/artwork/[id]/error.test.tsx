import { render, screen, fireEvent } from "@testing-library/react";

import Error from "./error";

// Mock window.location ve window.history
const mockReload = jest.fn();
const mockBack = jest.fn();

beforeAll(() => {
  Object.defineProperty(window, "location", {
    value: { reload: mockReload },
    writable: true,
  });

  Object.defineProperty(window, "history", {
    value: { back: mockBack },
    writable: true,
  });

  // console.error'Ä± mockla
  jest.spyOn(console, "error").mockImplementation(() => {});
});

afterEach(() => {
  jest.clearAllMocks();
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe("Error Component", () => {
  // Error interface'ine uygun mock error
  const mockError: Error & { cause?: unknown } = {
    name: "TestError",
    message: "Test error message",
    cause: "Some cause", // opsiyonel cause
  };

  it("renders correctly with default error message", () => {
    render(<Error error={mockError} />);

    expect(screen.getByText("Oops! Something went wrong.")).toBeInTheDocument();
  });

  it("logs the error with cause when provided", () => {
    const errorWithCause = {
      ...mockError,
      cause: { reason: "Test reason" },
    };

    render(<Error error={errorWithCause} />);

    expect(console.error).toHaveBeenCalledWith(errorWithCause);
  });

  it("logs the error without cause when not provided", () => {
    const errorWithoutCause = {
      name: "TestError",
      message: "Test error message",
    };

    render(<Error error={errorWithoutCause} />);

    expect(console.error).toHaveBeenCalledWith(errorWithoutCause);
  });

  it("calls window.location.reload when 'Try Again' button is clicked", () => {
    render(<Error error={mockError} />);
    fireEvent.click(screen.getByText("Try Again"));
    expect(mockReload).toHaveBeenCalledTimes(1);
  });

  it("calls window.history.back when 'Go Back' button is clicked", () => {
    render(<Error error={mockError} />);
    fireEvent.click(screen.getByText("Go Back"));
    expect(mockBack).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Error error={mockError} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
