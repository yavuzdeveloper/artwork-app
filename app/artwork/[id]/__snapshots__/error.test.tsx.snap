// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen flex flex-col items-center justify-center bg-gray-50 p-6"
  >
    <div
      class="max-w-md w-full text-center"
    >
      <h1
        class="text-3xl font-bold text-gray-800 mb-6"
      >
        Oops! Something went wrong.
      </h1>
      <div
        class="flex justify-center gap-4"
      >
        <button
          class="px-6 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition"
        >
          Try Again
        </button>
        <button
          class="px-6 py-2 bg-gray-300 text-gray-800 rounded-md hover:bg-gray-400 transition"
        >
          Go Back
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
